datafusion/core/src/physical_optimizer/update_aggr_exprs.rs

```diff
@@ -0,0 +1,185 @@
+// under the License.
+
+use std::sync::Arc;
+
+use super::PhysicalOptimizerRule;
+use datafusion_common::{
+    config::ConfigOptions,
+    tree_node::{Transformed, TransformedResult, TreeNode},
+};
+use datafusion_common::{plan_datafusion_err, Result};
+use datafusion_physical_expr::{
+    reverse_order_bys, AggregateExpr, EquivalenceProperties, PhysicalSortRequirement,
+};
+use datafusion_physical_plan::aggregates::concat_slices;
+use datafusion_physical_plan::windows::get_ordered_partition_by_indices;
+use datafusion_physical_plan::{
+    aggregates::AggregateExec, ExecutionPlan, ExecutionPlanProperties,
+};
+
+/// The optimizer rule check the ordering requirements of the aggregate expressions.
+/// There are 3 kinds of aggregators in terms of ordering requirement
+/// - `AggregateOrderSensitivity::Insensitive`
+/// - `AggregateOrderSensitivity::HardRequirement`
+/// - `AggregateOrderSensitivity::Beneficial`
+///
+/// `AggregateOrderSensitivity::Beneficial` mode have an ordering requirement. However, aggregator can still produce
+/// correct result even when ordering requirement is not satisfied (less efficiently). This rule analyzes
+/// `AggregateOrderSensitivity::Beneficial` aggregate expressions to see whether their requirement is satisfied or not.
+/// Using this information aggregators are updated to either work in efficient mode or less efficient mode.
+#[derive(Default)]
+pub struct OptimizeAggregateOrder {}
+
+impl OptimizeAggregateOrder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl PhysicalOptimizerRule for OptimizeAggregateOrder {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        plan.transform_up(&update_aggregator_when_beneficial).data()
+    }
+
+    fn name(&self) -> &str {
+        "OptimizeAggregateOrder"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+/// Updates the aggregators with mode `AggregateOrderSensitivity::Beneficial`
+/// if existing ordering enables to execute them more efficiently.
+fn update_aggregator_when_beneficial(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    if let Some(aggr_exec) = plan.as_any().downcast_ref::<AggregateExec>() {
+        if !aggr_exec.mode().is_first_stage() {
+            return Ok(Transformed::no(plan));
+        }
+        let input = aggr_exec.input();
+        let mut aggr_expr = aggr_exec.aggr_expr().to_vec();
+        let group_by = aggr_exec.group_by();
+
+        let input_eq_properties = input.equivalence_properties();
+        let groupby_exprs = group_by.input_exprs();
+        // If existing ordering satisfies a prefix of the GROUP BY expressions,
+        // prefix requirements with this section. In this case, aggregation will
+        // work more efficiently.
+        let indices = get_ordered_partition_by_indices(&groupby_exprs, input);
```

Comment:
`get_ordered_partition_by_indices()` is used here for group_by exprs. Perhaps the function name should be generalized.